### PR TITLE
AWS distributed training tutorial

### PR DESCRIPTION
**Pull request by inkawhich**
ATTN: @orionr @jspisak 
Initial commit of AWS distributed training tutorial. Right now this PR is just adding source code into beginner_source and I did not add it into the toc tree.

### CODE REVIEW

**Comment by JoelMarcey** [comment]
Is this something we want to merge in now?

**Comment by orionr** [comment]
Once someone has walked through it and we've confirmed the builds pass I think so.

**Comment by jspisak** [comment]
@brianjo - is taking a pass. Let’s hold off until he’s done and we figure out if this needs anything additional.

**Comment by jspisak** [comment]
@teng-li - is this something you can look at? We'd like to have an updated distributed tutorial ready for stable release.

cc @pietern

**Comment by teng-li** [comment]
@jspisak  Yes, on my TODO list

File: beginner_source/aws_distributed_training_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+import torch.utils.data
+import torch.utils.data.distributed
+import torchvision.transforms as transforms
+import torchvision.datasets as datasets
+import torchvision.models as models
+from torch.multiprocessing import Pool, Process, set_start_method
+try:
+    set_start_method('spawn')
+except RuntimeError:
+    pass
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: Please raise the error here instead of swallowing it.

File: beginner_source/aws_distributed_training_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+import torch.nn.parallel
+import torch.distributed as dist
+import torch.optim
+import torch.utils.data
+import torch.utils.data.distributed
+import torchvision.transforms as transforms
+import torchvision.datasets as datasets
+import torchvision.models as models
+from torch.multiprocessing import Pool, Process, set_start_method
+try:
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: Are you sure you don't need?

```
if __name__ = '__main__':
    set_start_method('spawn')
```

File: beginner_source/aws_distributed_training_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+# Imports
+# ~~~~~~~
+# 
+# The important distributed training specific imports here are
+# `torch.nn.parallel <https://pytorch.org/docs/stable/nn.html#torch.nn.parallel.DistributedDataParallel>`__,
+# `torch.distributed <https://pytorch.org/docs/stable/distributed.html>`__,
+# `torch.utils.data.distributed <https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler>`__,
+# and
+# `torch.multiprocessing <https://pytorch.org/docs/stable/multiprocessing.html>`__.
+# It is also important to set the multiprocessing start method to *spawn*,
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: or *fork_server* if you use Python 3

File: beginner_source/aws_distributed_training_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+                  'Prec@5 {top5.val:.3f} ({top5.avg:.3f})'.format(
+                   epoch, i, len(train_loader), batch_time=batch_time,
+                   data_time=data_time, loss=losses, top1=top1, top5=top5))
+
+def adjust_learning_rate(initial_lr, optimizer, epoch):
+    """Sets the learning rate to the initial LR decayed by 10 every 30 epochs"""
+    lr = initial_lr * (0.1 ** (epoch // 30))
+    for param_group in optimizer.param_groups:
+        param_group['lr'] = lr
+
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: This isn't enough to achieve the right accuracy, please refer to the learning schedule here.

https://github.com/pytorch/examples/pull/262/files

File: beginner_source/aws_distributed_training_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,675 @@
...
+# interesting part. Here is where we will define the inputs for the run.
+# Some of the inputs are standard model training inputs such as batch size
+# and number of training epochs, and some are specific to our distributed
+# training task. The required inputs are:
+# 
+# -  **batch\_size** - batch size for *each* process in the distributed
+#    training group. Total batch size across distributed model is
+#    batch\_size\*world\_size
+# 
+# -  **workers** - number of worker threads used with the dataloaders in
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: number of worker processes, not threads

File: beginner_source/aws_distributed_training_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+# several inputs. First, a *backend* input which specifies the backend to
+# use (i.e. NCCL, Gloo, MPI, etc.). An *init\_method* input which is
+# either a url containing the address and port of the rank0 machine or a
+# path to a non-existant file on the shared file system. Note, to use the
+# file init\_method, all machines must have access to the file, similarly
+# for the url method, all machines must be able to communicate on the
+# network so make sure to configure any firewalls and network settings to
+# accomodate. The *init\_process\_group* function also takes *rank* and
+# *world\_size* arguments which specify the rank of this process when run
+# and the number of processes in the collective, respectively.
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: Please also mention the env:// init method, which will use TCP init anyway.

File: beginner_source/aws_distributed_training_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,675 @@
...
+# multi-gpu nodes.
+# 
+
+print("Initialize Process Group...")
+# Initialize Process Group
+# v1 - init with url
+dist.init_process_group(backend=dist_backend, init_method=dist_url, rank=int(sys.argv[1]), world_size=world_size)
+# v2 - init with file
+#dist.init_process_group(backend="nccl", init_method="file:///home/ubuntu/pt-distributed-tutorial/trainfile", rank=int(sys.argv[1]), world_size=world_size)
+
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: v3 - init with env://, and please provide what the environmental variables should look like in the comment too.

File: beginner_source/aws_distributed_training_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,675 @@
...
+# The next major step is to initialize the model to be trained. Here, we
+# will use a resnet18 model from ``torchvision.models`` but any model may
+# be used. First, we initialize the model and place it in GPU memory.
+# Next, we make the model ``DistributedDataParallel``, which handles the
+# distribution of the data to and from the model and is critical for
+# distributed training. The ``DistributedDataParallel`` module also
+# handles the averaging of gradients across the world, so we do not have
+# to explicitly average the gradients in the training step.
+# 
+# It is important to note that this is a blocking function, meaning
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: which function is a blocking function?

File: beginner_source/aws_distributed_training_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,675 @@
...
+transform = transforms.Compose(
+    [transforms.Resize(224),
+     transforms.ToTensor(),
+     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+
+# Initialize Datasets. STL10 will automatically download if not present
+trainset = datasets.STL10(root='./data', split='train', download=True, transform=transform)
+valset = datasets.STL10(root='./data', split='test', download=True, transform=transform)
+
+# Create DistributedSampler to handle distributing the dataset across nodes when training
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: Also add a comment saying this needs to be called after torch.init_process_group is called

File: beginner_source/aws_distributed_training_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+    # train for one epoch
+    print("\nBegin Training Epoch {}".format(epoch+1))
+    train(train_loader, model, criterion, optimizer, epoch)
+
+    # evaluate on validation set
+    print("Begin Validation @ Epoch {}".format(epoch+1))
+    prec1 = validate(val_loader, model, criterion)
+
+    # remember best prec@1 and save checkpoint if desired
+    #is_best = prec1 > best_prec1
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: do you need to keep this code and comment here?

File: beginner_source/aws_distributed_training_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,675 @@
...
+# line arguments, respectively. Once uploaded, open two ssh terminals into
+# each node.
+# 
+# -  On the first terminal for node0, run ``$ python main.py 0 0``
+# 
+# -  On the second terminal for node0 run ``$ python main.py 1 1``
+# 
+# -  On the first terminal for node1, run ``$ python main.py 2 0``
+# 
+# -  On the second terminal for node1 run ``$ python main.py 3 1``
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: I would prefer you add an argument parser for the main.py and avoid using these 0 0  1 1 arguments. They should come with a name to be more clear to the user which number means what.

File: beginner_source/aws_distributed_training_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,675 @@
...
+# envrionment. If you are using AWS, don't forget to **SHUT DOWN YOUR
+# NODES** if you are not using them or you may find an uncomfortably large
+# bill at the end of the month.
+# 
+# **Where to go next**
+# 
+# -  Check out the `launcher
+#    utility <https://pytorch.org/docs/stable/distributed.html#launch-utility>`__
+#    for a different way of kicking off the run
+#
```
Comment (teng-li) [inline beginner_source/aws_distributed_training_tutorial.py]: If you use python 3.4 or above, check out the ImageNet example in pytorch example. This will use `torch.multiprocessing.spawn`. 

Please also mention about this for another way to launch multiprocesses. Please also take a look at the latest pytorch example image net code, which has this implemented

**Comment by orionr** [comment]
@teng-li Nate is no longer contracting. Would you mind taking over this PR and making the appropriate changes? If you have issues, let me know and we can do it together. Thanks.

**Comment by teng-li** [comment]
@orionr  I am making those changes, will send a new PR soon, stay tuned

**Comment by teng-li** [comment]
Let's move here: https://github.com/pytorch/tutorials/pull/368/checks
